source/Types.cpp

```diff
@@ -72,31 +72,59 @@ namespace Aws
             return {};
         }
 
-        String Base64Encode(const Vector<uint8_t> &encode)
+        String Base64Encode(const Vector<uint8_t> &encode) noexcept
         {
-            ByteCursor toEncode = aws_byte_cursor_from_array((const void *)encode.data(), encode.size());
+            auto toEncode = aws_byte_cursor_from_array((const void *)encode.data(), encode.size());
 
-            size_t allocation_size = 0;
-
-            if (aws_base64_compute_encoded_len(encode.size(), &allocation_size) == AWS_OP_SUCCESS)
+            size_t allocationSize = 0;
+            if (aws_base64_compute_encoded_len(toEncode.len, &allocationSize) == AWS_OP_SUCCESS)
             {
-                String output(allocation_size, 0x00);
-                ByteBuf tempBuf = aws_byte_buf_from_array(output.data(), output.size());
-                tempBuf.len = 0;
+                String outputStr(allocationSize, 0x00);
+                auto tempBuf = aws_byte_buf_from_empty_array(outputStr.data(), outputStr.capacity());
+                UnsafeInteropHelpers::Base64Encode(toEncode, tempBuf);
 
-                if (aws_base64_encode(&toEncode, &tempBuf) == AWS_OP_SUCCESS)
+                // encoding appends a null terminator, and accounts for it in the encoded length,
+                // which makes the string 1 character too long
+                if (tempBuf.len == allocationSize - 1)
                 {
-                    // encoding appends a null terminator, and accounts for it in the encoded length,
-                    // which makes the string 1 character too long
-                    if (output.back() == 0)
+                    outputStr.pop_back();
+                    return outputStr;
+                }
+            }
+
+            return {};
+        }
+
+        namespace UnsafeInteropHelpers
+        {
+            void Base64Decode(const ByteCursor &toDecode, ByteBuf &out) noexcept
```

Comment:
so hopefully i understand this, but i think this is a helper function that can be summarized as "dont attempt to decode if we dont have the space for it, if we do, decode into the buffer". Do we want to expose this as a public API or have a be a impl detail of the the existing API for Base64, leaning towards it as a impl detail of the existing API.

Reply:
i rolled this back for now. im not against having those helpers in general. my major concern is that underlying crt decode impl always starts decoding into output buffer at 0, while the binding assumes it would start appending to output buffer (which it probably should) and we should not expose that foot gun until we fix it. 

second concern is the naming. calling them unsafe, while there are 30 other places where we do the same thing for crypto, seems very inconsistent. maybe we can do something better where we can avoid allocating memory and also avoid unsafeness of byte buf